File: buildbot_step.py
#! /usr/bin/env python

import datetime
import distutils.spawn
import math
import os
import subprocess
import sys
import time

HELP = """
Invalid number of arguments ...

Usage:
    buildbot_step <title> <command...>

Examples:
    buildbot_step 'List my directory the long way' ls -lah
""".strip()

BB_LINE = "========= {} ========="

TIME_FORMAT = '%Y-%m-%d %H:%M:%S.%f'

def status_line(type, step, code, duration):
    line = "{} {} (results: {}, elapsed: {} secs) (at {})"
    print(BB_LINE.format(line.format(
        type,
        step,
        code,
        int(math.ceil(duration)),
        datetime.datetime.now().strftime(TIME_FORMAT)
    )))



def main(args):
    """
    Run the given command with buildbot specific output the first argument is
    always the name of the command...
    """

    if len(args) < 2 or args[0] == '-h':
        print(HELP)

    title = args[0]
    command = args[1:]

    # Hack around the fact that call does not know how to handle relative
    # executables...
    executable = os.path.abspath(distutils.spawn.find_executable(command[0]))
    command[0] = executable

    start = time.time()
    status_line(title, 'Started', 0, 0)
    exit_code = subprocess.call(command)
    status_line(title, 'Finished', exit_code, time.time() - start)
    sys.exit(exit_code)


main(sys.argv[1:])

File: buildbot_step.js
#! /usr/bin/env node

var spawn = require('child_process').spawn;
var argv = process.argv;
var format = require('util').format;

var BB_LINE = '========= %s =========';

function padZero(input, pad) {
  pad = pad || 2;
  input = String(input)
  if (input.length < pad) {
    return '0' + input;
  }
  return input;
}

function ensureLength(input, len) {
  input = String(input);
  if (input.length < len) {
    var missing = len - input.length;
    for (var i = 0; i < missing; i++) {
      input = input + '0';
    }
    return input;
  }

  if (input.length > len) {
    return input.slice(0, 6);
  }

  return input;
}

function status(type, step, code, duration) {
  var line = '%s %s (results: %s, elapsed: %d secs) (at %s)'
  var date = new Date();

  // Time shown in UTC to be the least confusing...
  var dateFormat = format(
    '%s-%s-%s %s:%s:%s.%s',
    date.getUTCFullYear(),
    padZero(date.getMonth()),
    padZero(date.getDate()),
    padZero(date.getHours()),
    padZero(date.getMinutes()),
    padZero(date.getSeconds()),
    ensureLength(date.getMilliseconds(), 6)
  );
  console.log(format(BB_LINE, format(
    line,
    type,
    step,
    code,
    Math.ceil(duration),
    dateFormat
  )))
}

var title = argv[3];
var command = argv.slice(3);

if (!title || !command.length) {
  console.error(format(
    'Error missing title or command\n ' +
    'Usage: %s <title> <command...>',
    argv[1]
  ));
  process.exit(1);
}

var start = Date.now();

status('Started', argv.slice(2).join(' '), 0, 0);
var proc = spawn(argv[3], argv.slice(4), {
  stdio: 'inherit'
});

proc.once('exit', function(code) {
  status('Finished', argv.slice(2).join(' '), code, (Date.now() - start) / 1000);
  process.exit(code);
});
